torch/optim/adam.py

```diff
@@ -16,19 +16,24 @@ class Adam(Optimizer):
             running averages of gradient and its square (default: (0.9, 0.999))
         eps (float, optional): term added to the denominator to improve
             numerical stability (default: 1e-8)
-        weight_decay (float, optional): weight decay (L2 penalty) (default: 0)
+        weight_decay (float, optional): weight decay factor (default: 0)
+        l2_reg (boolean, optional): whether to use the original L2
+            weight regularization or the weight decay method from the paper
+            `Fixing Weight Decay Regularization in Adam` (default: True)
```

Comment:
Maybe it would be worth to even issue a `FutureWarning` stating that `l2_reg` will be set to `False` by default in a future version?

```python
def __init__(self, params, lr=1e-3, betas=(0.9, 0.999), eps=1e-8,
             weight_decay=0, l2_reg="warn", amsgrad=False):
    if l2_reg == "warn":
        warnings.warn("l2_reg will be set to False by default in a future Pytorch version."
                      " Set l2_reg=False explicit to silence this warning.", FutureWarning)
        l2_reg = True  # backward compatibility
    ...
```

Reply:
Thanks @ogrisel for your kind suggestions, I appreciate you taking time to do this. I have fixed the name of the paper everywhere. 

Regarding the recommendation of the default value of the `l2_reg` flag, I am happy to change it but it depends on the PyTorch maintainers... I suspect they are busy with other stuff, but hopefully they reply, this issue has been open now for a year so I am not too hopeful 😿 

If you have time can you also review the changes in the schedular, since now if we use this method, the learning rate and weight decay variables are decoupled and I need to schedule them both and that is a bit messy... Any suggestions to simplify that aspect will be great! Thanks!

Reply:
With respect to getting reviewers attention, I believe it would be worth taking the time to try to reproduce some of the experiments of the AdamW paper in gist and then publish a screenshot of the train loss and validation accuracy convergence plots in a comments in this PR (or even include those in the description of the PR itself).